packages/swapper/src/thorchain-utils/getL1RateOrQuote.ts

```diff
@@ -0,0 +1,430 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import { CHAIN_NAMESPACE, fromAssetId } from '@shapeshiftoss/caip'
+import * as adapters from '@shapeshiftoss/chain-adapters'
+import {
+  assertUnreachable,
+  bn,
+  bnOrZero,
+  convertDecimalPercentageToBasisPoints,
+  convertPrecision,
+  fromBaseUnit,
+  isFulfilled,
+  isRejected,
+  toBaseUnit,
+} from '@shapeshiftoss/utils'
+import type { Result } from '@sniptt/monads'
+import { Err, Ok } from '@sniptt/monads'
+import { v4 as uuid } from 'uuid'
+
+import { getDefaultSlippageDecimalPercentageForSwapper } from '../index'
+import type {
+  CommonTradeQuoteInput,
+  GetEvmTradeQuoteInput,
+  GetEvmTradeRateInput,
+  GetTradeRateInput,
+  GetUtxoTradeRateInput,
+  ProtocolFee,
+  QuoteFeeData,
+  SwapErrorRight,
+  SwapperDeps,
+} from '../types'
+import { SwapperName, TradeQuoteError } from '../types'
+import { makeSwapErrorRight } from '../utils'
+import * as evm from './evm'
+import { getLimitWithManualSlippage } from './getLimitWithManualSlippage/getLimitWithManualSlippage'
+import { getQuote } from './getQuote'
+import { addLimitToMemo, getAffiliate, getNativePrecision, getSwapSource } from './index'
+import type {
+  ThorEvmTradeQuote,
+  ThorEvmTradeRate,
+  ThornodeQuoteResponseSuccess,
+  ThorTradeQuote,
+  ThorTradeRate,
+  ThorTradeRoute,
+  ThorTradeUtxoOrCosmosQuote,
+  ThorTradeUtxoOrCosmosRate,
+} from './types'
+import { TradeType } from './types'
+import * as utxo from './utxo'
+
+const SAFE_GAS_LIMIT = '100000' // depositWithExpiry()
+
+type ThorTradeRateOrQuote = ThorTradeRate | ThorTradeQuote
+type ThorEvmTradeRateOrQuote = ThorEvmTradeRate | ThorEvmTradeQuote
+type ThorUtxoOrCosmosTradeRateOrQuote = ThorTradeUtxoOrCosmosRate | ThorTradeUtxoOrCosmosQuote
+
+type MakeThorTradeInputBase = {
+  route: ThorTradeRoute
+  memo: string
+  allowanceContract: string
+  feeData: QuoteFeeData
+}
+
+type MakeThorTradeInput<T extends ThorTradeRateOrQuote> = T extends ThorEvmTradeRateOrQuote
+  ? MakeThorTradeInputBase & {
+      data: string
+      router: string
+      vault: string
+    }
+  : MakeThorTradeInputBase & {
+      data?: never
+      router?: never
+      vault?: never
+    }
+
+export const getL1RateOrQuote = async <T extends ThorTradeRateOrQuote>(
+  input: T extends ThorTradeRate ? GetTradeRateInput : CommonTradeQuoteInput,
+  deps: SwapperDeps,
+  streamingInterval: number,
+  swapperName: SwapperName,
+): Promise<Result<T[], SwapErrorRight>> => {
+  const {
+    sellAsset,
+    buyAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmountCryptoBaseUnit,
+    accountNumber,
+    receiveAddress,
+    affiliateBps: requestedAffiliateBps,
+  } = input
+
+  const {
+    config,
+    assertGetEvmChainAdapter,
+    assertGetUtxoChainAdapter,
+    assertGetCosmosSdkChainAdapter,
+  } = deps
+
+  const nativePrecision = getNativePrecision(swapperName)
```

Comment:
nitpick but feels like we should be able to access the native `Asset` somehow instead of this, though with swapper dependency injection, meh, may not be so easy, feel free to disregard me

Reply:
And to add, it isn't necessarily the native asset === native precision. In the case of Mayachain, CACAO is the native asset and the only asset that uses precision 10, everything else is precision 8 (which is updated in the final pr in this stack)